File: Stack/1249.Minimum_Remove_to_Make_Valid_Parentheses/Minimum_Remove_to_Make_Valid_Parentheses.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        
        arr = list(s)
        
        stk = []
        
        for i, curr in enumerate(s):

            if curr == "(":
                stk.append(i)
                
            if curr == ")":
                if stk != []: stk.pop()
                else: arr[i] = ""
                    
            
                
                    
                    
        for i in stk: arr[i] = ""
        
        return "".join(arr)

#solution 2
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        LeftToRight = []
        l = 0
        for i in s:
            if i == "(":
                l += 1
                LeftToRight.append(i)
            elif i == ")":
                if l > 0:
                    l -= 1
                    LeftToRight.append(i)
            else: LeftToRight.append(i)
                
        RightToLeft = []
        r = 0
        for i in LeftToRight[::-1]:
            if i == ")":
                r += 1
                RightToLeft.append(i)
            elif i == "(":
                if r > 0:
                    r -= 1
                    RightToLeft.append(i)
            else: RightToLeft.append(i)
                
        return "".join(RightToLeft[::-1])




            

File: Stack/1249.Minimum_Remove_to_Make_Valid_Parentheses/Minimum_Remove_to_Make_Valid_Parentheses.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        vector<int> stk;
        string ans;
        for(int i = 0; i < s.size();i++){
            if(s[i] == '(') stk.push_back(i);
            else if(s[i] == ')'){
                if(stk.size() == 0){
                    s[i] = '$';
                }
                else{
                    stk.pop_back();
                }
            }
        }
        
        for(int i = 0; i < stk.size(); i++){
            s[stk[i]] = '$';
        }
        
        for(char i: s){
            if(i != '$') ans += i;
        }
        
        return ans;
        
    }
};